graphix/pattern.py

```diff
@@ -1080,6 +1166,370 @@ def to_qasm3(self, filename):
+
+    def shift_signals(self):
+        """Shift signals to the back based on signal destinations."""
+        self.collect_signal_destination()
+        for node_index in self.morder + self.output_nodes:
+            signal = self.nodes[node_index].Mprop[3]
+            self.nodes[node_index].Mprop[3] = []
+            for signal_label, destinated_nodes in self.signal_destination[node_index].items():
+                for destinated_node in destinated_nodes:
+                    node = self.nodes[destinated_node]
+                    if signal_label == "Ms":
+                        node.Mprop[2] += signal
+                    elif signal_label == "Mt":
+                        node.Mprop[3] += signal
+                    elif signal_label == "X":
+                        node.Xsignal += signal
+                    elif signal_label == "Z":
+                        node.Zsignal += signal
+                    else:
+                        raise ValueError(f"Invalid signal label: {signal_label}")
+
+    def get_graph(self):
+        """Get a graph from a local pattern
+
+        Returns
+        -------
+        nodes : list
+            list of node indices
+        edges : list
+            list of edges
+        """
+        nodes = []
+        edges = []
+        for index, node in self.nodes.items():
+            nodes.append(index)
+            for cmd in node.seq:
+                if cmd >= 0:
+                    if index > cmd:
+                        edges.append((cmd, index))
+        return nodes, edges
+
+    def get_pattern(self):
+        """Convert a local pattern into a corresponding global pattern. Currently, only standardized pattern is supported.
+
+        Returns
+        -------
+        pattern : Pattern
+            standardized global pattern
+        """
+        pattern = Pattern(output_nodes=self.output_nodes)
+        Nseq = [["N", i] for i in self.nodes.keys()]
+        Eseq = []
+        Mseq = []
+        Xseq = []
+        Zseq = []
+        Cseq = []
+        for node_index in self.morder + self.output_nodes:
+            node = self.nodes[node_index]
+            for cmd in node.seq:
+                if cmd >= 0:
+                    Eseq.append(node.get_command(cmd))
+                    self.nodes[cmd].seq.remove(node_index)
+                elif cmd == -1:
+                    Mseq.append(node.get_command(cmd))
+                elif cmd == -2:
+                    Xseq.append(node.get_command(cmd))
+                elif cmd == -3:
+                    Zseq.append(node.get_command(cmd))
+                elif cmd == -4:
+                    Cseq.append(node.get_command(cmd))
+                else:
+                    raise ValueError(f"command {cmd} is invalid!")
+            if node.result is not None:
+                pattern.results[node.index] = node.result
+        pattern.seq = Nseq + Eseq + Mseq + Xseq + Zseq + Cseq
+        return pattern
+
+
+def xor_combination_list(list1, list2):
```

Comment:
Can we use python set operations?

Reply:
I think it will be complicated to use set because it cannot distinguish one where a signal appears only once from where a signal appears more than once. (e.g. set treats the following as the same, [1]. [1, 1], [1, 1, 1] -> {1}).
Do you have any idea to solve the above problem?